specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -0,0 +1,1409 @@
+    /// add it to the calling process' package graph. Once the dependency has
+    /// been added other code-loading methods (LoadLibrary, CoCreateInstance, etc)
+    /// can find the binaries in the resolved package.
+    ///
+    /// Package resoution is specific to a user. The same package dependency can
+    /// resolve to different packages for different users on a system.
+    ///
+    /// This adds the resulting package to the process' package graph.
+    /// A process' package graph is used to search for DLLs (per Dynamic-Link Library Search Order),
+    /// WinRT objects and other resources; the caller can now load DLLs, activate
+    /// WinRT objects and use other resources from the framework package until
+    /// PackageDependencyContext.Close() is called (or the process ends).
+    /// The package dependency Id must match a package dependency defined
+    /// for the calling user or the system (via PinForSystem) or an exception is raised.
+    ///
+    /// Each successful call adds the resolve packaged to the
+    /// calling process' package graph, even if already present. There is no
+    /// duplicate 'detection' or 'filtering' applied by the API (multiple
+    /// references to a package is not harmful). Once resolution is complete
+    /// the package stays resolved for that user until the last reference across
+    /// all processes for that user is removed via PackageDependencyContext.Close()
+    /// (or process termination).
+    ///
+    /// This adds the resolved package to the caller's package graph, per rank.
+    /// A process' package graph is a list of packages sorted by rank in ascending
+    /// order (-infinity...0...+infinity). If package(s) are present in the
+    /// package graph with the same rankrank the resolved package is
+    /// (by default) added after others of the same rank. To add a package
+    /// before others of the same rank, specify PackageDependency.PrependIfRankCollision.
+    ///
+    /// Calls to Add() can be balanced by a PackageDependencyContext.Close() (or object destruction)
+    /// to remove the entry from the package graph.
+    PackageDependencyContext Add(AddPackageDependencyOptions options);
+}
+
+/// A unique identifier for a resolved package dependency
+struct PackageDependencyContextId
+{
+    UInt64 Id;
+};
+
+/// This object provides access to information about a package dependency context
+/// and removes it from the caller's package graph when closed (via .Close() / .Dispose())
+/// or destroyed (i.e. undo packageDependency.Add()).
+///
+/// Closing this object is the moral equivalent of FreeLibrary.
+///
+/// @note This does not unload loaded resources (DLLs etc). After removing
+///        a package dependency any files loaded from the package can continue
+///        to be used; future package dependency resolution (via new calls to
+///        PackageDependency.Add) will fail to see the removed package dependency.
+runtimeclass PackageDependencyContext : ICloseable
```

Comment:
`Context` is a weird term. Isn't this a `ResolvedPackageReference`? And the `Add` method should be called `Resolve`.

Reply:
DISCUSSION: Add maps to a particular package, does bookkeeping to make the package active, and returns a {thing}. This object represents the fact that the package has been added to your process graph.

DISCUSSION: Make sure the little sequence diagram shows how these are generated and managed.